inbm/cloudadapter-agent/proto/inbc-mock/inbc-mock.go

```diff
@@ -0,0 +1,114 @@
+package main
+
+import (
+	"log"
+	"net"
+	"time"
+
+	pb "inbc-mock/pb"
+
+	"github.com/google/uuid"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+// server is used to implement inbs.INBSSBServiceServer.
+type server struct {
+	pb.UnimplementedINBSSBServiceServer
+}
+
+func authStreamInterceptor(srv interface{}, stream grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+	md, ok := metadata.FromIncomingContext(stream.Context())
+	if !ok {
+		log.Print("Missing metadata from client")
+		return status.Errorf(codes.InvalidArgument, "Missing metadata")
+	}
+
+	token, ok := md["token"]
+	if !ok || len(token) == 0 || token[0] != "good_token" { // Replace with your expected token.
+		log.Print("Invalid token given by client: " + token[0])
+		return status.Errorf(codes.Unauthenticated, "invalid token")
+	}
+
+	// If the metadata is valid, proceed with handling the stream
+	return handler(srv, stream)
+}
+
+func (s *server) INBMCommand(stream pb.INBSSBService_INBMCommandServer) error {
+	ctx := stream.Context()
+
+	// Extract metadata from the context
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		log.Fatalf("Failed to retrieve metadata")
+		return status.Errorf(codes.InvalidArgument, "Missing metadata")
+	}
+
+	// Retrieve the inband-id from metadata
+	nodeIDVals, ok := md["node-id"]
+	if !ok || len(nodeIDVals) == 0 {
+		log.Fatalf("node-id is required in metadata")
+		return status.Errorf(codes.Unauthenticated, "node-id is required")
+	}
+	nodeID := nodeIDVals[0] // Using the first value
+
+	log.Printf("Received connection with inband-id: %s", nodeID)
+
+	// Loop until the context is done (connection closed)
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			// Sending a Ping to the client
+			requestId := uuid.New().String()
+			err := stream.Send(&pb.INBMRequest{
+				RequestId: requestId,
+				Payload: &pb.INBMRequest_PingRequest{
+					PingRequest: &pb.PingRequest{},
+				},
+			})
+			if err != nil {
+				log.Fatalf("Failed to send a ping: %v", err)
```

Comment:
**<h2>$\color{#15F4EE}Miscellaneous$</h2>**
**Suggestion:** Replace log.Fatalf with log.Printf followed by a return statement in the INBMCommand method to prevent the server from terminating upon encountering an error. This change will enhance the robustness of the server by allowing it to continue operating even if an individual stream encounters an issue.
**NOTE:** Ensure that the suggested code is comprehensive. This serves as an example of the code you could use. The user should verify it after accepting the suggestion.
```suggestion
   if err != nil {
       log.Printf("Failed to send a ping: %v", err)
       return err
   }
```

Reply:
this is a test client